docs/design/model-reader-writer-implementation-design.md

```diff
@@ -0,0 +1,244 @@
+# Azure SDK Design Spec for implementation of `ModelReaderWriter`
+
+- [Summary](#summary)
+- [Problem Statement or Scenario Description](#problem-statement-or-scenario-description)
+- [Scope: Goals and Non-Goals](#scope-goals-and-non-goals)
+- [Feature Design](#feature-design)
+- [Detailed Implementation](#detailed-implementation)
+
+## Summary
+
+This design specification is for the implementation of `ModelReaderWriter` related interfaces on the generated libraries, not the design specification of `ModelReaderWriter` itself.
+
+The implementation of `ModelReaderWriter` might change in the future when we have further consolidations of existing interfaces.
+
+For details of `ModelReaderWriter`, please refer to [ModelReaderWriter](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/README.md).
+
+## Problem Statement or Scenario Description
+
+Our SDK libraries ship with a set of generated models which are used to represent the data structures of the service. These models are shipped with a set of serialization and deserialization but those are only designed for us to use internally by our generated operation to build the payload or parse the response back to the models.
+
+In this design, our customers will not be able to serialize or deserialize the models in the same way as we do. They could only serialize the model using the default way which is not usually correct. An intuitive example of this is when there is an "extensible enum" type in the model:
+```csharp
+public partial class Foo
+{
+    public ExtensibleEnumType? Bar { get; set; }
+}
+
+public partial struct ExtensibleEnumType : IEquatable<ExtensibleEnumType>
+{
+    private readonly string _value;
+
+    public ExtensibleEnumType(string value)
+    {
+        _value = value ?? throw new ArgumentNullException(nameof(value));
+    }
+}
+```
+
+The default behavior of the most widely used JSON serializer and deserializer is usually: they serialize classes and structs in the same way, assuming that they are models containing properties. This is true for our classes, but for structs, it is not always true. The extensible enums are modeled as structs, but they are actually serialized to a string value in our internal implementation when building the payload or parsing the response. For example, the `Foo` model above are expected to be serialized to the following JSON:
+```json
+{
+    "bar": "value"
+}
+```
+but in the default behavior of `System.Text.JsonSerializer`, it will be serialized to:
```

Comment:
We need mention that the `ModelReaderWriter` does not support `System.Text.JsonSerializer` now, so even we implement it in .Net SDK but still can't serialize the model properly by using `System.Text.JsonSerializer`.

Reply:
added a paragraph about this